src/hal/motor.cpp

```diff
@@ -2,20 +2,461 @@
+    DEBUG_PRINT("state: MOTOR_STATE_OFF");
+
+    motor_pwm_disable();
+    digitalWrite(PIN_MOTOR_ENABLE, LOW);
+    motor_busy = false;
+  } else if (next_state == MOTOR_STATE_HOLD) {
+    DEBUG_PRINT("state: MOTOR_STATE_HOLD");
+
+    motor_pwm_disable();
+    digitalWrite(PIN_MOTOR_ENABLE, HIGH);
+    motor_busy = false;
+  } else if (next_state == MOTOR_STATE_OPEN) {
+    DEBUG_PRINT("state: MOTOR_STATE_OPEN");
+
+    // TODO
+    /*
+    if (motor_state == MOTOR_STATE_CLOSE) {
+      // Should going directly from the CLOSE state to the OPEN state with no
+      // intermediate state transition constitute an error?
+      //
+      // Not sure if restoring interrupt state is necessary here.
+      SREG = sreg; // restore interrupt state
+
+      assert(motor_state != MOTOR_STATE_CLOSE);
+    }
+    */
+
+    digitalWrite(PIN_MOTOR_DIRECTION, PIN_MOTOR_DIRECTION_OPEN);
+    digitalWrite(PIN_MOTOR_ENABLE, HIGH);
+    motor_pwm_enable();
+  } else if (next_state == MOTOR_STATE_CLOSE) {
+    DEBUG_PRINT("state: MOTOR_STATE_CLOSE");
+
+    // TODO
+    /*
+    if (motor_state == MOTOR_STATE_OPEN) {
+      // Should going directly from the OPEN state to the CLOSE state with no
+      // intermediate state transition be an error?
+
+      // Not sure if restoring interrupt state is necessary here.
+      SREG = sreg; restore interrupt state
+
+      assert(motor_state != MOTOR_STATE_OPEN);
+    }
+    */
+
+    digitalWrite(PIN_MOTOR_DIRECTION, PIN_MOTOR_DIRECTION_CLOSE);
+    digitalWrite(PIN_MOTOR_ENABLE, HIGH);
+    motor_pwm_enable();
+  }
+
+  motor_state = next_state;
+
+  SREG = sreg; // restore interrupt state
+}
+
+//*****************************************************************************
+// Motor Position Control
+//*****************************************************************************
+
+// Set the current position as the zero point.
+void motor_position_zero()
+{
+  assert((motor_state != MOTOR_STATE_OPEN) && (motor_state != MOTOR_STATE_CLOSE));
+
+  uint8_t sreg = SREG; // save interrupt state
+  cli(); // disable interrupts
+
+  motor_control.microstep_position = 0;
+  motor_control.step_position = 0;
+  motor_control.step_command = 0;
+
+  SREG = sreg; // restore interrupt state
+}
+
+// Moves the motor to an angle relative to its current position.
+//
+// Parameters
+// ----------
+// int32_t angle      The angle (in degrees) to move the motor.
+// int8_t  direction  The direction to move the motor. One of:
+//                      * MOTOR_DIRECTION_OPEN
+//                      * MOTOR_DIRECTION_CLOSE
+void motor_move(int32_t angle, int8_t direction)
+{
+  // TODO: Do we want to have assertions? How are they implemented in this
+  // system? (i.e. Do they cause a watchdog reset, print a debug message to the
+  // console, etc.)
+  assert((direction == MOTOR_DIRECTION_OPEN) ||
+         (direction == MOTOR_DIRECTION_CLOSE));
+
+  uint8_t sreg = SREG; // save interrupt state
+  cli(); // disable interrupts
```

Comment:
See above on atomic blocks

Reply:
Switched to new syntax for atomic blocks.